File: src/amr_controller/amr_controller/joy_to_twist.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Joy
from geometry_msgs.msg import Twist

class JoyToTwist(Node):
    def __init__(self):
        super().__init__('joy_to_twist')
        self.declare_parameter('linear_axis', 1)
        self.declare_parameter('angular_axis', 3)
        self.declare_parameter('deadman_button', 0)
        self.declare_parameter('linear_scale', 0.5)
        self.declare_parameter('angular_scale', 1.0)

        self.linear_axis    = self.get_parameter('linear_axis').value
        self.angular_axis   = self.get_parameter('angular_axis').value
        self.deadman_button = self.get_parameter('deadman_button').value
        self.linear_scale   = self.get_parameter('linear_scale').value
        self.angular_scale  = self.get_parameter('angular_scale').value

        self.pub = self.create_publisher(Twist, 'joy_vel', 10)
        self.sub = self.create_subscription(Joy, 'joy', self.joy_cb, 10)
        self.get_logger().info('joy_to_twist node started')

    def joy_cb(self, msg: Joy):
        if msg.buttons[self.deadman_button] != 1:
            return
        twist = Twist()
        twist.linear.x  = msg.axes[self.linear_axis]  * self.linear_scale
        twist.angular.z = msg.axes[self.angular_axis] * self.angular_scale
        self.pub.publish(twist)

def main(args=None):
    rclpy.init(args=args)
    node = JoyToTwist()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/amr_controller/src/joy_to_twist.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/joy.hpp"
#include "geometry_msgs/msg/twist.hpp"

class JoyToTwist : public rclcpp::Node
{
public:
  JoyToTwist() : Node("joy_to_twist"),
    linear_axis_(3),
    angular_axis_(0),
    deadman_axis_1_(4),
    deadman_axis_2_(5), 
    linear_scale_(1.0),
    angular_scale_(1.0)
  {
    pub_ = this->create_publisher<geometry_msgs::msg::Twist>("joy_vel", 10);
    sub_ = this->create_subscription<sensor_msgs::msg::Joy>(
      "joy", 10,
      std::bind(&JoyToTwist::joy_cb, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "joy_to_twist node started");
  }

private:
  void joy_cb(const sensor_msgs::msg::Joy::SharedPtr msg)
  {
    bool deadman_pressed = (msg->axes[deadman_axis_1_] == -1.0) ||
                           (msg->axes[deadman_axis_2_] == -1.0);

    if (!deadman_pressed) return;

    geometry_msgs::msg::Twist twist;
    twist.linear.x  = msg->axes[linear_axis_]  * linear_scale_;
    twist.angular.z = msg->axes[angular_axis_] * angular_scale_;
    pub_->publish(twist);
  }

  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_;
  rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr sub_;

  const int    linear_axis_, angular_axis_, deadman_axis_1_, deadman_axis_2_;
  const double linear_scale_, angular_scale_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<JoyToTwist>());
  rclcpp::shutdown();
  return 0;
}